Working Python 2.7 support under Microsoft Windows

clach04 opened the pull request:
* Added a bare bones readme
* Use the newer version of libpebble2 (0.0.28) from Rebble GitHub repo, rather than what appears to be unmaintained PyPi version (appears because the links in the page at https://pypi.org/project/libpebble2/ are to the old unmaintained repo, not the current https://github.com/pebble-dev/pebble-tool/ one)

Tested with:

    Python 2.7.10 (default, May 23 2015, 09:40:32) [MSC v.1500 32 bit (Intel)] on win32

Ready for 2nd review 2023-07-22 (12:11 Pacific time)

clach04 committed:
Working Python 2.7 support under Microsoft Windows

Tested with:

    Python 2.7.10 (default, May 23 2015, 09:40:32) [MSC v.1500 32 bit (Intel)] on win32

clach04 commented:
This was enough so I could get logs out of the running watchface/app on my phone.

I think pyreadline is OK to include, https://pypi.org/project/pyreadline/ claims to support Python3.

https://pypi.org/project/anyreadline/ might be another alternative but I only have experience with pyreadline  hence using it.

clach04 commented:
@Katharine anything else this would need to be considered mergeable? You did the last commit so picked you to ping :-p

clach04 commented:
@ishotjr do you have time and permission to merge this? There are a few other PRs that would be nice to get consolidated

As always thanks to everyone on the Rebble time, you all rock!

Katharine commented on requirements.txt:
```diff
@@ -1,4 +1,5 @@
-libpebble2==0.0.26
+#libpebble2==0.0.28
+git+https://github.com/pebble-dev/libpebble2.git@685cd9841e647b48609ce904a87540019390b3be#egg=libpebble2
```
The libpebble2 on pypi [isn't unmaintained](https://pypi.org/project/libpebble2/) - I'd rather use that.

clach04 replied:
Ahhh, that's not clear from the pypi page. All the links are to the old pebble project and CI. Thanks for making that clear :-). I was wondering maybe  I can open a PR to clean that up, but the readme at https://github.com/pebble-dev/libpebble2/blob/master/README.rst doesn't match the (long) description over at https://pypi.org/project/libpebble2/ - I'll get a bug opened over there but notes (mostly to myself)

https://pypi.org/project/libpebble2/ points to:

* https://github.com/pebble/libpebble2 (multiple places) - instead of https://github.com/pebble-dev/libpebble2
* https://travis-ci.org/pebble/libpebble2 - which also points to https://github.com/pebble/libpebble2

I don't remember why I felt the direct link was necessary, I'm reasonably confident I was ending up with the wrong/old version and it didn't work without it.

clach04 replied:
Issues are not enabled on https://github.com/pebble-dev/libpebble2 - https://pypi.org/project/libpebble2/ looks abandoned (readme/description and links are not the current readme in https://github.com/pebble-dev/libpebble2). This may be a pypi issue?

clach04 replied:
...
> I don't remember why I felt the direct link was necessary, I'm reasonably confident I was ending up with the wrong/old version and it didn't work without it.

lol, I missed the setup.py requirements - that's probably why :-D

Katharine commented on requirements.txt:
```diff
@@ -16,3 +17,4 @@ websocket-client==0.32.0
 wheel==0.24.0
 colorama==0.3.3
 packaging==16.7
+pyreadline==2.1
```
A quick look suggests that pyreadline has been unmaintained for longer than this tool, and does not work under the latest version of Python. pyreadline3 might be a better choice, though it's very obscure.

Does installing this without importing it do anything?

clach04 replied:
I'll need to get back to you on this, been a while since I looked at it.

clach04 replied:
> A quick look suggests that pyreadline has been unmaintained for longer than this tool, and does not work under the latest version of Python. pyreadline3 might be a better choice, though it's very obscure.

Do you mean https://pypi.org/project/pyreadline3 ? That doesn't look appropriate, it does NOT support Python 2.7 :-(, only some 3.x

> Does installing this without importing it do anything?

Are you asking if the install runs some code in the background I'd need to look at the setup.py to know? I'm not 100% sure what you are asking. It's defintely being imported by commands/repl.py https://github.com/pebble-dev/pebble-tool/blob/92561f5c075fe5c812b69409ef338a3548928472/pebble_tool/commands/repl.py#L5

Katharine replied:
Supporting Python 2.7 is not really a goal at this point, given that Python 2.7 has been dead for years. That said, it is still necessary given that while this runs (or should run) fine under Python 3, the build chain does not. (For that matter, supporting Windows outside WSL is also a non-goal...)

Is editing requirements.txt even the right thing to be doing? Selectively installing things might be better. What, if anything, happens if this is installed on Linux or macOS?

clach04 replied:
Hmmm, maybe that's the way to tackle this. Scratch this PR, and I open a new one that's ONLY a readme change with a note for py2.7 (and Windows) so it's a manual step (or a separate requirements file) so the current known platforms work the way they always have.

Let me try that out real quick.

Katharine replied:
It should be possible to selectively install things in setup.py, if one uses that mechanism instead of requirements.txt.

clach04 replied:
I'm definitely down for that :-)

I just posted a hopefully easier to accept PR https://github.com/pebble-dev/pebble-tool/pull/4 to fix things on all platforms, with Windows to follow. I already tested out requirements windows file and it works great, but a conditional in setup.py would be great too. My main goal is to get something merged so it's in a central place, for a clean working version.

Katharine commented on README.md:
```diff
@@ -0,0 +1,8 @@
+# Pebble build tool
```
It's the Pebble Tool, not the Pebble Build Tool.

clach04 replied:
I'll fix that now get it done. If things look good I can squash changes into a single commit.

Thanks for taking the time for review!

Katharine reviewed (changes requested):
My number one question would be: why not just run under WSL? You might get the pure python part working, which I'm not opposed to, but most of the functionality will fail awkwardly. It's designed for WSL.

clach04 commented:
> My number one question would be: why not just run under WSL? You might get the pure python part working, which I'm not opposed to, but most of the functionality will fail awkwardly. It's designed for WSL.

Not sure if this is a reasonable answer but building a native binary (pebble.exe) ;-) ... and I don't have to get to grips with WSL ;-p the different versions originally confused me. Seems to have settled down with WSL2.

clach04 committed:
Fix two thirds of review comments

This commit is untested.

clach04 committed:
Fix issues with setup.py - possibly Windows specific

Match versions in setup.py with (working) requirements.txt.

When ran from setup.py httplib2 would pull in pyparsing 3.1.0
rather than 2.4.7. This does not happen with pip install of
requirements.txt, manually added pyparsing with pinned version number.

clach04 committed:
Include pyreadline in setup.py to match requirements.txt

clach04 commented:
@Katharine and @ishotjr I think I'm ready for round 2 :-)

Thanks again for you help with this.

clach04 commented:
> @Katharine and @ishotjr I think I'm ready for round 2 :-)
> 
> Thanks again for you help with this.

I was writing this when https://github.com/pebble-dev/pebble-tool/pull/2#discussion_r1271333431 was posted, hold fire on review. New approach being tested.

clach04 commented:
Replaced by:

* https://github.com/pebble-dev/pebble-tool/pull/4
* https://github.com/pebble-dev/pebble-tool/pull/5
